### PR TITLE
feat: [IDP-167]: API support to fetch & store status

### PR DESCRIPTION
**Pull request by vigneswara-propelo**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)

### CODE REVIEW

**Comment by vigneswara-propelo** [comment]
trigger messagecheck

**Comment by vigneswara-propelo** [comment]
trigger messagecheck

**Review by vikyathharekal — COMMENTED** [review]
> Reviewed 14 of 21 files at r1, 3 of 5 files at r2, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 23 file(s) to review. 0 of 2 approvals obtained. 

___
*[`idp-service/service/src/main/java/io/harness/idp/app/IdpPersistenceConfig.java` line 55 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOP27bPCuzUYIZW3Oz2:-NOP27bPCuzUYIZW3Oz3:b8wkz5g) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/app/IdpPersistenceConfig.java#L55)):*
> ```Java
> @GuiceModule
> @EnableMongoRepositories(basePackages = {"io.harness.idp.config.repositories", "io.harness.idp.secret.repositories",
>                              "io.harness.idp.namespace.repositories", "io.harness.idp.status.repositories"},
> ```

It would be good if we could have a wildcard for this. "io.harness.idp.*.repositories"



_Code quote:_
```Java
status
```
___
*[`idp-service/service/src/main/java/io/harness/idp/app/IdpPersistenceConfig.java` line 58 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOP2PgF14H00wJaIynW:-NOP2PgF14H00wJaIynX:bek9a9t) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/app/IdpPersistenceConfig.java#L58)):*
> ```Java
>     includeFilters = @ComponentScan.Filter(HarnessRepo.class), mongoTemplateRef = "primary")
> @EnableMongoAuditing
> @OwnedBy(HarnessTeam.PL)
> ```

IDP. Just noticed. Not related to this PR.



_Code quote:_
```Java
PL
```
___
*[`idp-service/service/src/main/java/io/harness/idp/status/mappers/StatusInfoMapper.java` line 30 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOP5PJe42CXkL1yygI6:-NOP5PJe42CXkL1yygI7:b5klv8k) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/mappers/StatusInfoMapper.java#L30)):*
> ```Java
> 
>   public StatusInfoEntity fromDTO(StatusInfo statusInfo) {
>     return StatusInfoEntity.builder()
> ```

What about type? UI needs to know the type as well right?
___
*[`idp-service/service/src/main/java/io/harness/idp/status/repositories/StatusInfoRepositoryCustomImpl.java` line 52 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOPDgva29uqtKcKaii8:-NOPDgva29uqtKcKaii9:b-4tigbw) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/repositories/StatusInfoRepositoryCustomImpl.java#L52)):*
> ```Java
>     update.set(StatusInfoEntityKeys.status, statusInfoEntity.getStatus());
>     update.set(StatusInfoEntityKeys.reason, statusInfoEntity.getReason());
>     if (statusInfoEntity.getLastModifiedAt() == null) {
> ```

Will this be null ever? I guess this field will be auto-populated as we have annotated it with @LastModifiedDate
And its not something that would need to be updated 
___
*[`idp-service/service/src/main/java/io/harness/idp/status/resources/StatusInfoApiImpl.java` line 12 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOPEGyU2Sb2ocW62GBB:-NOPEGyU2Sb2ocW62GBC:b-igb837) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/resources/StatusInfoApiImpl.java#L12)):*
> ```Java
> import io.harness.annotations.dev.HarnessTeam;
> import io.harness.annotations.dev.OwnedBy;
> import io.harness.idp.status.beans.StatusInfoEntity;
> ```

API Impl need not know about Entity. It should deal with DTOs
___
*[`idp-service/service/src/main/java/io/harness/idp/status/resources/StatusInfoApiImpl.java` line 38 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOPEimN-ntl-ezj_CcM:-NOPEimN-ntl-ezj_CcN:bqe7qac) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/resources/StatusInfoApiImpl.java#L38)):*
> ```Java
>     StatusInfoResponse statusResponse = new StatusInfoResponse();
>     if (statusInfo.isEmpty()) {
>       log.warn("Could not fetch status for accountId {}", harnessAccount);
> ```

Should we log the type as well?
___
*[`idp-service/service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java` line 38 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOPEz9X4V1wzLyQhsjP:-NOPEz9X4V1wzLyQhsjQ:bhh887v) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/service/StatusInfoServiceImpl.java#L38)):*
> ```Java
>   @Override
>   public StatusInfo save(StatusInfo statusInfo, String accountIdentifier, String type) {
>     StatusInfoEntity statusInfoEntity = StatusInfoMapper.fromDTO(statusInfo);
> ```

nit: we could have accountIdentifier as a param in fromDTO and the next line could be taken care of internally.
___
*[`idp-service/service/src/main/java/io/harness/idp/status/enums/TypeEnum.java` line 10 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOP4BmR5ZcODUR-ugxT:-NOP4BmR5ZcODUR-ugxU:b-cbkidv) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/enums/TypeEnum.java#L10)):*
> ```Java
> package io.harness.idp.status.enums;
> 
> public enum TypeEnum { INFRA, ONBOARDING }
> ```

Do we need an Enum suffix? Or could we just name them as Status & StatusType



_Code quote:_
```Java
TypeEnum
```
___
*[`idp-service/service/src/test/java/io/harness/idp/status/resources/StatusInfoApiImplTest.java` line 76 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOPGA3DEUPfZI3FL4dx:-NOPGA3DEUPfZI3FL4dy:b-a0hp37) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/test/java/io/harness/idp/status/resources/StatusInfoApiImplTest.java#L76)):*
> ```Java
> 
>   void initializeStatusInfo() {
>     statusInfo = new StatusInfo();
> ```

nit: Isn't it better to create an instance locally and return it? The class variable could be modified by other tests which could side effects

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 23 file(s) to review. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/service/src/main/java/io/harness/idp/status/mappers/StatusInfoMapper.java` line 30 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOP5PJe42CXkL1yygI6:-NOPMxez6qUrqJAB7tIX:b-chzu2k) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/mappers/StatusInfoMapper.java#L30)):*
> What about type? UI needs to know the type as well right?
UI will anyways send type in the URL /v1/status-info/{type}. Hence, thought it wont be required.
Let me know your thoughts if we need to send type in response
___
*[`idp-service/service/src/main/java/io/harness/idp/status/repositories/StatusInfoRepositoryCustomImpl.java` line 52 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOPDgva29uqtKcKaii8:-NOPOVSsC30cWZmrMXR7:bolqv3i) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/repositories/StatusInfoRepositoryCustomImpl.java#L52)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Will this be null ever? I guess this field will be auto-populated as we have annotated it with @LastModifiedDate
And its not something that would need to be updated 
</blockquote></details>

when I was testing, this field is getting auto-populated only the first time we create the document.
if we try to update the same doc, the latest time is not updated there.
Also, in case of infra status we could use timestamp from pods as the DB update time might slightly differ

File: idp-service/contracts/openapi/SNAPSHOT/openapi.yaml [hunk 1]
```diff
@@ -92,6 +92,42 @@ paths:
       responses:
         '204':
           description: No Content
+  '/v1/status-info/{type}':
```
Comment (sandeepa-kv-jois) [inline idp-service/contracts/openapi/SNAPSHOT/openapi.yaml:95]: can we have a better naming of these endpoints to convey what they mean for

**Review by vikyathharekal — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 23 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/service/src/main/java/io/harness/idp/status/mappers/StatusInfoMapper.java` line 30 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOP5PJe42CXkL1yygI6:-NOPWnVB4V6DtEF5gnun:b-5wza4m) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/mappers/StatusInfoMapper.java#L30)):*
<details><summary><i>Previously, vigneswara-propelo (Vigneswara) wrote…</i></summary><blockquote>

> What about type? UI needs to know the type as well right?
UI will anyways send type in the URL /v1/status-info/{type}. Hence, thought it wont be required.
Let me know your thoughts if we need to send type in response
</blockquote></details>

Cool. Should be fine for now then.

File: idp-service/contracts/openapi/SNAPSHOT/openapi.yaml (hunk 1 above)
Comment (sandeepa-kv-jois) [reply idp-service/contracts/openapi/SNAPSHOT/openapi.yaml:95]: May be onboardingStatus?

Comment (vigneswara-propelo) [reply idp-service/contracts/openapi/SNAPSHOT/openapi.yaml:95]: But the other day we decided to have a common api for onboarding & Infra status right? 
Keeping it onboarding-status would become specific to onboarding alone.
UI can fetch status info based on type
1. Infra - /v1/status-info/infra
2. Onboarding - /v1/status-info/onboarding

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 23 file(s) to review. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/service/src/main/java/io/harness/idp/app/IdpPersistenceConfig.java` line 55 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOP27bPCuzUYIZW3Oz2:-NOPqBJj77GsXyotKHPh:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/app/IdpPersistenceConfig.java#L55)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

It would be good if we could have a wildcard for this. "io.harness.idp.*.repositories"
</blockquote></details>

Done.
___
*[`idp-service/service/src/main/java/io/harness/idp/app/IdpPersistenceConfig.java` line 58 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOP2PgF14H00wJaIynW:-NOPqBva-aGpL0GGud0D:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/app/IdpPersistenceConfig.java#L58)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

IDP. Just noticed. Not related to this PR.
</blockquote></details>

Done.
___
*[`idp-service/service/src/main/java/io/harness/idp/status/resources/StatusInfoApiImpl.java` line 12 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOPEGyU2Sb2ocW62GBB:-NOPqOcG8GFIKyQEw9d8:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/resources/StatusInfoApiImpl.java#L12)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

API Impl need not know about Entity. It should deal with DTOs
</blockquote></details>

Done.
___
*[`idp-service/service/src/main/java/io/harness/idp/status/resources/StatusInfoApiImpl.java` line 38 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOPEimN-ntl-ezj_CcM:-NOPqU5C4cj-j4yiHO3d:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/resources/StatusInfoApiImpl.java#L38)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Should we log the type as well?
</blockquote></details>

Done.
___
*[`idp-service/service/src/main/java/io/harness/idp/status/enums/TypeEnum.java` line 10 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOP4BmR5ZcODUR-ugxT:-NOPqGUj1WOh5WGUclgl:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/service/src/main/java/io/harness/idp/status/enums/TypeEnum.java#L10)):*
<details><summary><i>Previously, vikyathharekal (Vikyath Harekal) wrote…</i></summary><blockquote>

Do we need an Enum suffix? Or could we just name them as Status & StatusType
</blockquote></details>

Done.

File: idp-service/contracts/openapi/SNAPSHOT/openapi.yaml (hunk 1 above)
Comment (sandeepa-kv-jois) [reply idp-service/contracts/openapi/SNAPSHOT/openapi.yaml:95]: Resolving as it's a common api for onboarding & Infra status. So, we can use status-info.

**Review by deepharness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 23 file(s) to review. 1 change requests. 6. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/service/src/main/java/io/harness/idp/status/repositories/StatusInfoRepositoryCustomImpl.java` line 27 at r4](https://reviewable.io/reviews/harness/harness-core/43995#-NOQhaZh4_LXIiaivwd1:-NOQhaZh4_LXIiaivwd2:b-fbqfbo) ([raw file](https://github.com/harness/harness-core/blob/bc2a06489cd734fc912f911400ea5bd7f24e7b43/idp-service/service/src/main/java/io/harness/idp/status/repositories/StatusInfoRepositoryCustomImpl.java#L27)):*
> ```Java
> 
>   @Override
>   public StatusInfoEntity saveOrUpdateStatusInfo(StatusInfoEntity statusInfoEntity) {
> ```

saveOrUpdateStatusInfo should be saveOrUpdate as its a part of StatusInfoRepo

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 23 file(s) to review. 1 change requests. 7. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`idp-service/service/src/main/java/io/harness/idp/status/repositories/StatusInfoRepositoryCustomImpl.java` line 27 at r4](https://reviewable.io/reviews/harness/harness-core/43995#-NOQhaZh4_LXIiaivwd1:-NOSIS01-NaXwTILAjav:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/bc2a06489cd734fc912f911400ea5bd7f24e7b43/idp-service/service/src/main/java/io/harness/idp/status/repositories/StatusInfoRepositoryCustomImpl.java#L27)):*
<details><summary><i>Previously, deepharness wrote…</i></summary><blockquote>

saveOrUpdateStatusInfo should be saveOrUpdate as its a part of StatusInfoRepo
</blockquote></details>

Done.

**Review by vikyathharekal — APPROVED** [review]
> Reviewed 7 of 11 files at r4, 5 of 5 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 23 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by vikyathharekal — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 23 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by sandeepa-kv-jois — APPROVED** [review]

**Review by sandeepa-kv-jois — APPROVED** [review]
> Reviewed 7 of 21 files at r1, 2 of 5 files at r2, 1 of 6 files at r3, 7 of 11 files at r4, 5 of 5 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 1 change requests. 1. unresolved discussions.

**Review by vigneswara-propelo — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 1 change requests. 2. unresolved discussions. 

___
*[`idp-service/contracts/openapi/SNAPSHOT/openapi.yaml` line 95 at r3](https://reviewable.io/reviews/harness/harness-core/43995#-NOJr5xgzjMUnLziV72o-r3-95:-NOgzaDW8vdPPuoYSbxz:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/d2dae720035e252cd1c6c366c6ebebd7d0a6d023/idp-service/contracts/openapi/SNAPSHOT/openapi.yaml#L95)):*
<details><summary><i>Previously, sandeepa-kv-jois wrote…</i></summary><blockquote>

Resolving as it's a common api for onboarding & Infra status. So, we can use status-info.
</blockquote></details>

Done.

**Review by sandeepa-kv-jois — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: 1 change requests. 1. unresolved discussions.

**Review by deepharness — APPROVED** [review]
> Reviewed 7 of 21 files at r1, 2 of 5 files at r2, 1 of 6 files at r3, 7 of 11 files at r4, 5 of 5 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43995)* status: :shipit: complete! ✓

**Pull request merged**